clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -784,6 +785,17 @@ void SemaHLSL::handleSV_DispatchThreadIDAttr(Decl *D, const ParsedAttr &AL) {
                  HLSLSV_DispatchThreadIDAttr(getASTContext(), AL));
 }
 
+void SemaHLSL::handleSV_GroupIDAttr(Decl *D, const ParsedAttr &AL) {
+  auto *VD = cast<ValueDecl>(D);
+  if (!isLegalTypeForHLSLSV_ThreadOrGroupID(VD->getType())) {
+    Diag(AL.getLoc(), diag::err_hlsl_attr_invalid_type)
+        << AL << "uint/uint2/uint3";
```

Comment:
Done. Moved the duplicate diagnostic code into `isLegalTypeForHLSLSV_ThreadOrGroupID` function.

Still use `isLegalTypeForHLSLSV_ThreadOrGroupID` as function name, because it seems only Group or Thread ID support `uint/unit2/uint3` type.

Reply:
There are 4 that have the same validation requirement:

* SV_DispatchThreadID
* SV_GroupThreadID
* SV_GroupID
* SV_DispatchGrid

All of them except the last one have `ID` in the name, which is why I suggested that naming. I think having the name of th method be too specific will be misleading as we extend to support additional semantics.

I also think this pattern of 3-element vectors for IDs/indices is super common, so we may do more of this in the future.

Reply:
Also, it is important that the function be named for what it does, so if you moved the diagnostic into the function it should be called `diagnose...` rather than `is...`.

Reply:
@llvm-beanz  Done. Renamed the function name to `diagnoseInputIDType`